geometry/sat_collision.h

```diff
@@ -0,0 +1,251 @@
+     * @return vertex that belongs to this shape and is closest to @point
+     * @see Shape::GetClosestVertex
+     */
+    Vec2 GetClosestVertex(Vec2 point) const override;
+
+    /**
+     * @brief Implements @Shape.Project for a circle.
+     * @param axis to project @this shape on
+     * @param minRes smallest value of the resulting line
+     * @param maxRes largest value of the resulting line
+     * @return void
+     * @see Shape::Project
+     */
+    void Project(Vec2 axis, double& minRes, double& maxRes) const override;
+ private:
+    std::vector<Vec2> m_Vertices;
+};
+
+/**
+ * @brief Circle representation. Consists of a centre (position vector) and a
+ *        radius
+ */
+class Circle : public Shape {
+ public:
+    Circle(Vec2 centre, double radius);
+
+    /**
+     * @brief implements @Shape.SAT for a circle.
+     * @param otherShape Shape pair to test overlapping against
+     * @return true if there is no gap on any of the projections.
+     *         false otherwise.
+     * @see Shape::SAT
+     */
+    bool SAT(const Shape& otherShape) const override;
+
+    /**
+     * @brief Implements @Shape.GetClosestVertex for a circle.
+     * @param point to compare vertices with, represented by a position vector
+     * @return Always returns the centre of the circle, since this is the only
+     *         point (vertex) stored by the shape.
+     * @see Shape::GetClosestVertex
+     */
+    Vec2 GetClosestVertex(Vec2 point) const override;
+
+    /**
+     * @brief Implements @Shape.Project for a circle.
+     * @param axis to project @this shape on
+     * @param minRes smallest value of the resulting line
+     * @param maxRes largest value of the resulting line
+     * @return void
+     * @see Shape::Project
+     */
+    void Project(Vec2 axis, double& minRes, double& maxRes) const override;
+ private:
+    double m_Radius;
+    Vec2 m_Centre;
+};
+
+/**
+ * @brief Find if two intervals [min1, max1] and [min2, max2] overlaps
+ * @param min1 min point of line segment 1
+ * @param max1 max point of line segment 1
+ * @param min2 min point of line segment 2
+ * @param max2 max point of line segment 2
+ * @return `true` if the line segments overlap
+ * @return `false` otherwise
+ */
+bool DoOverlap(double min1, double max1, double min2, double max2);
+
+/**
+ * @brief Find if two shapes overlap when projected onto an axis
+ * @param axis Axis to project shapes onto
+ * @param shape1 First shape
+ * @param shape2 Second shape
+ * @return `true` if the two polygons overlap when projected onto the axis.
+ * @return `false` otherwise
+ */
+bool OverlapOnAxis(const Vec2& axis, const Shape& shape1,
+              const Shape& shape2);
+
+/**
+ * @brief Performs the SAT collision algorithm on two 2-dimensional shapes
+ * @param shape1 First shape
+ * @param shape2 Second shape
+ * @return `true` if the shapes collide.
+ * @return `false` otherwise
+ */
+bool sat(const Shape& shape1, const Shape& shape2);
+}}
```

Comment:
```suggestion
}  // namespace collision
}  // namespace geometry

```


__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix&nbsp; | &nbsp;**Tags:** Inelegant